File: sorting/IntermediateSort/mergeSort.py
def merge(list1, list2):
    i = j = 0
    result = []
    while(i < len(list1) and j < len(list2)):
        if(list1[i] < list2[j]):
            result.append(list1[i])
            i += 1
        else:
            result.append(list2[j])
            j += 1
        
    while i < len(list1):
        result.append(list1[i])
        i += 1
    while j < len(list2):
        result.append(list2[j])
        j += 1
   
    return result

def mergeSort(list1):
    
    if len(list1) <= 1:
        return list1
    
    middle = round(len(list1) / 2)
    left = mergeSort(list1[:middle])
    
    right = mergeSort(list1[middle:])

    return merge(left, right)



print(mergeSort([2,1,5,3]))

File: sorting/IntermediateSort/MergeSort.js
'use strict'

// const cutoff = (arr,start = 0, end = arr.length -1) =>{
//     const result = [];
//     for(let i = start; i <= end; i++){
//         result.push(arr[i])
//     }
//     return result;
// }

// console.log(cutoff([1,2,3,4,5,6,7,9],1,2));


const mergeSort = arr =>{
    if (arr.length <= 1) return arr;
    const mid = Math.floor(arr.length / 2)
    const left = arr.slice(0,mid);
    const right = arr.slice(mid);

    return merge(mergeSort(left), mergeSort(right))
}



const merge = (arr1,arr2) =>{

    let left = 0;
    let right = 0;
    const result = [];

    while(left < arr1.length && right < arr2.length){

        if(arr1[left] < arr2[right]){
            result.push(arr1[left]);
            left++
        }
        else{
            result.push(arr2[right])
            right++
        }
    }
    return result.concat(arr2.slice(right), arr1.slice(left));
}

console.log(mergeSort([13,12,24,23]));